api/src/main/java/io/opencensus/metrics/DoubleGaugeMetric.java

```diff
@@ -0,0 +1,237 @@
+ * Copyright 2018, OpenCensus Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.metrics;
+
+import io.opencensus.common.ToDoubleFunction;
+import io.opencensus.internal.Utils;
+import java.util.List;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Double Gauge metric, to report instantaneous measurement of a double value. Gauges can go both up
+ * and down. The gauges values can be negative. For example, Free memory or Total memory. See {@link
+ * io.opencensus.metrics.MetricRegistry} for an example of its use.
+ *
+ * <p>Example 1: Create a Gauge without a labels
+ *
+ * <pre>{@code
+ * class YourClass {
+ *
+ *   private static final MetricRegistry metricRegistry = Metrics.getMetricRegistry();
+ *   DoubleGaugeMetric totalMemory = metricRegistry.addDoubleGaugeMetric(
+ *       "Total_memory", "Total CPU Memory", "1", new ArrayList<LabelKey>());
+ *
+ *   Point defaultPoint = totalMemory.getDefaultPoint();
+ *
+ *   void doWork() {
+ *      defaultPoint.inc();
+ *      // Your code here.
+ *      defaultPoint.dec();
+ *   }
+ * }
+ *
+ * }</pre>
+ *
+ * <p>Example 2: You can also use labels(keys and values) to track different types of metric.
+ *
+ * <pre>{@code
+ * class YourClass {
+ *
+ *   private static final MetricRegistry metricRegistry = Metrics.getMetricRegistry();
+ *
+ *   List<LabelKey> keys = Collections.singletonList(LabelKey.create("Type","desc"));
+ *   DoubleGaugeMetric totalMemory = metricRegistry.addDoubleGaugeMetric(
+ *       "Total_Memory", "Total CPU Memory", "1", keys);
+ *
+ *   List<LabelValue> usedMemory = Collections.singletonList(LabelValue.create("Used"));
+ *   Point point = totalMemory.addPoint(usedMemory);
+ *
+ *   void doSomeWork() {
+ *      point.set(12.5);
+ *      // Your code here.
+ *      point.dec();
+ *   }
+ * }
+ * }</pre>
+ *
+ * @since 0.17
+ */
+@ThreadSafe
+public abstract class DoubleGaugeMetric {
+
+  /**
+   * Adds and returns a {@code Point}. This is more convenient form when you can want to manually
+   * increase and decrease values as per your service requirements. The number of label values must
+   * be the same to that of the label keys passed to {@link MetricRegistry#addDoubleGaugeMetric}.
+   *
+   * @param labelValues the list of label values.
+   * @return a {@code Point} the value of single gauge.
+   * @since 0.17
+   */
+  public abstract Point addPoint(List<LabelValue> labelValues);
```

Comment:
We have changed the function name from addPoint to addTimeSeries. 

> The caller has to keep several variables or a map to lookup points again based on label values. That adds a lot of code overhead and essentially just replicates what the metric has internally already.
Giving this add-or-get semantics seems preferable. If performance is a concern, the caller can still keep a point directly in a variable.

I agree with your concern, but for self sufficient gauge (with obj and function) we do not return Point object as users can do whatever they want. I think one option is to expose 
```Point getPoint(List<LabelValue> labels) ``` function which should work only with manually operating gauge, otherwise throw an exception.